### DISCUSSION TITLE
WLAN functions do not work from core1

### DISCUSSION THREAD

**Discussion by ayjaym** [micropython/micropython#11950] [RP2040 / Pico]
I've posted to raspberrypi.stackexchange.com with the details but in essence WLAN functionality doesn't appear to work when invoked from core1. (tested with Micropython 1.20.0). For example, configuring an AP hangs when the AP is enabled, but the exact same code works perfectly from core0. This is so even if the code is guarded by a lock. This is a bit of a showstopper as I had implemented rtpMIDI and want the message handling loop running in core1 so it can run concurrently with other activities on core0.

If you configure the AP from core0 and then set up a poll object bound to two UDP ports (for the rtpMIDI protocol) then passing the poll 
object to core1 causes the poll to correctly timeout but no UDP datagrams are received. That code run entirely from core0 works fine. Again, guarding with a lock doesn't resolve the problem.

https://raspberrypi.stackexchange.com/questions/143640/unable-to-perform-any-wlan-related-operations-from-core1-on-pi-pico-with-micropy

**Comment by scruss** [comment]
I don't know if it's a hardware issue or older library problem, but other platforms running on the Raspberry Pi Pico W have reported only being able to access wifi functions from one core (ref: [WiFi (Raspberry Pi Pico W) Support — Arduino-Pico 3.3.0 documentation](https://arduino-pico.readthedocs.io/en/latest/wifi.html#important-information)).

Can your project be implemented under asyncio?

**Reply by ayjaym** [reply]
Thanks Stewart. It appears from that link that wlan functions can't be called across cores but not that they are expected to work only on core0. 
My project is a midi controller (it will be released as full hardware and software open source) which is intended to be user programmable in python. 
It has a 16x8 matrix of buttons suitable for a sequencer plus microsd and an oled display.
The case and buttons etc are 3d printed. It costs about £70 to build.
The hardware is all working perfectly.
I wanted to include midi over usb ( it has midi out via a 3 pin trs jack already) but Micropython doesn't yet have stable functionality in the main branch and circuit python, which does, has no support for multiple cores. 
So I then implemented a simple python class to do rtpmidi and this works with the free rtpmidi driver available for windows. But I wanted the midi traffic to be handled on core1 as rtpmidi is quite chatty with regular clock syncs, and then leave core0 for the users code. So asyncio doesn't really seem like the answer here.
I may try either rewriting the wlan stuff in c and seeing if that works on core1 but my suspicion is that this may be an SDK issue and not a micropython one.
Option 2 is to try wrapping tinyusb in an .mpy and running that on core1 with a python callback when midi is received and a queue for outbound midi. That would give me wired midi over usb.
To be honest I was surprised there weren't more mature micropython solutions for midi but I know it's being worked on. The Pico is the perfect controller for this project as the button matrix is actually implemented using a low cost resistive touch screen over an array of ws2812 LEDs and the Pico ADCs make it easy to read the screen directly without additional hardware.
